src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs

```diff
@@ -322,20 +339,45 @@ protected void OnUnloaded(VisualElement frameworkElement, Action action)
 			frameworkElement.Unloaded += unloaded;
 		}
 
-		protected Task OnUnloadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
+		protected async Task OnUnloadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
 		{
-			timeOut = timeOut ?? TimeSpan.FromSeconds(2);
 			TaskCompletionSource<object> taskCompletionSource = new TaskCompletionSource<object>();
-			OnUnloaded(frameworkElement, () => taskCompletionSource.SetResult(true));
-			return taskCompletionSource.Task.WaitAsync(timeOut.Value);
+			try
+			{
+				timeOut = timeOut ?? TimeSpan.FromSeconds(2);
+				OnUnloaded(frameworkElement, () => taskCompletionSource.TrySetResult(true));
+				await taskCompletionSource.Task.WaitAsync(timeOut.Value);
+			}
+			catch (TimeoutException)
+			{
+				if (!frameworkElement.IsLoadedOnPlatform() && !frameworkElement.IsLoaded)
+				{
+					taskCompletionSource.TrySetResult(true);
+				}
+				else
+					throw;
+			}
 		}
 
-		protected Task OnLoadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
+		protected async Task OnLoadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
 		{
-			timeOut = timeOut ?? TimeSpan.FromSeconds(2);
 			TaskCompletionSource<object> taskCompletionSource = new TaskCompletionSource<object>();
-			OnLoaded(frameworkElement, () => taskCompletionSource.SetResult(true));
-			return taskCompletionSource.Task.WaitAsync(timeOut.Value);
+			try
+			{
+				timeOut = timeOut ?? TimeSpan.FromSeconds(2);
+				OnLoaded(frameworkElement, () => taskCompletionSource.TrySetResult(true));
+				await taskCompletionSource.Task.WaitAsync(timeOut.Value);
+			}
+			catch (TimeoutException)
+			{
+				if (frameworkElement.IsLoadedOnPlatform() &&
+					frameworkElement.IsLoaded)
+				{
+					taskCompletionSource.TrySetResult(true);
+				}
+				else
+					throw;
+			}
```

Comment:
Very very similar methods, same logic, only change is the delegate, I'm wondering if we could move that to a common method and then work with it, for example:
```csharp
protected async Task AsyncWrapper (VisualElement frameworkElement, Action<VisualElement, Action> syncAction, TimeSpan? timeOut = null)
{
	TaskCompletionSource<object> taskCompletionSource = new TaskCompletionSource<object>();
	try
 	{
 		timeOut = timeOut ?? TimeSpan.FromSeconds(2);
 		syncAction(frameworkElement, () => taskCompletionSource.TrySetResult(true));
 		await taskCompletionSource.Task.WaitAsync(timeOut.Value);
 	}
    catch (TimeoutException)
 	{
 		if (frameworkElement.IsLoadedOnPlatform() &&
 			frameworkElement.IsLoaded)
 		{
 			taskCompletionSource.TrySetResult(true);
 		}
 		else
 			throw;
 	}
}
protected async Task OnUnloadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
	=> AsyncWrapper(frameworkElement, OnUnloaded, timeOut);

protected async Task OnLoadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
	=> AsyncWrapper(frameworkElement, OnLoaded, timeOut);

```
That way, if we have a bug in the async call, we know we fixed it for both cases, and we have less code to work with.

Reply:
I am terrible at naming, so AsyncWrapper might not be the best name, but you get the idea.

Reply:
@PureWeen After you mentioned the diff catch condition:
```csharp
protected async Task AsyncWrapper (VisualElement frameworkElement, Action<VisualElement, Action> syncAction, Func<bool> onTimeOut, TimeSpan? timeOut = null)
{
	TaskCompletionSource<object> taskCompletionSource = new TaskCompletionSource<object>();
	try
 	{
 		timeOut = timeOut ?? TimeSpan.FromSeconds(2);
 		syncAction(frameworkElement, () => taskCompletionSource.TrySetResult(true));
 		await taskCompletionSource.Task.WaitAsync(timeOut.Value);
 	}
        catch (TimeoutException)
 	{
 		if (onTimeOut (taskCompletionSource)) {
 		 	taskCompletionSource.TrySetResult(true);
 	 	}
 	 	else
 		 	throw;
 	}
}

protected async Task OnUnloadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
       => AsyncWrapper(frameworkElement, 
              OnUnloaded, 
              () => !frameworkElement.IsLoadedOnPlatform() && !frameworkElement.IsLoaded), 
              timeOut);

protected async Task OnLoadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
       => AsyncWrapper(frameworkElement, 
              OnLoaded,  
              () => frameworkElement.IsLoadedOnPlatform() && frameworkElement.IsLoaded,
              timeOut);
```
As I said, my naming is terrible :) 

And now question, why dows the tis use object and not bool?

Reply:
Just deleted this code and consolidated into the two already existing methods. 

Personally, I'd rather just install Rx and FromEventPattern everything  :-)